File: sdks/java/reactor-sdk/src/main/java/inc/reactor/sdk/internal/Completions.java
package inc.reactor.sdk.internal;

import inc.reactor.sdk.ReactorException;
import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.MethodType;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;
import org.jspecify.annotations.Nullable;

/**
 * Async operations, from {@code reactor_completion_fn} to {@link CompletableFuture}.
 *
 * <p>One registry per client. Each operation gets an id, the id travels to the FFI as {@code
 * userdata}, and the completion finds its future by it — so nothing here holds a strong reference
 * to the public client, and a pending operation cannot keep a session alive.
 *
 * <p>Two rules this class exists to keep:
 *
 * <ul>
 *   <li><b>Decode before claiming the future.</b> A future settles once. Marking it done and then
 *       converting the payload means a field of the wrong type throws where nothing can be settled
 *       any more, and the caller gets a hang instead of the typed error they were promised.
 *   <li><b>A successful completion that will not parse is a decode failure.</b> Substituting an
 *       empty object would make a schema declaring nothing indistinguishable from a model declaring
 *       nothing. An <i>absent</i> payload is a different answer and still means "nothing to report".
 * </ul>
 */
public final class Completions {

    /** Turns a completion's {@code result_json} into what the caller asked for. */
    @FunctionalInterface
    public interface Decoder<T> {
        /**
         * @param resultJson the payload, or {@code null} when the operation reports none
         * @return the decoded value
         * @throws RuntimeException when the payload cannot be understood
         */
        T decode(@Nullable String resultJson);
    }

    /** Makes the native call for an operation, given what to hand it. */
    @FunctionalInterface
    public interface Invoker {
        /**
         * @param callback the completion function pointer
         * @param userdata what the completion finds this operation by
         */
        void call(MemorySegment callback, MemorySegment userdata);
    }

    /** What an operation needs to hand the FFI: the callback and the userdata that finds it again. */
    public record Ticket(MemorySegment callback, MemorySegment userdata) {}

    private record Pending<T>(String operation, Decoder<T> decoder, CompletableFuture<T> future) {}

    private static final MethodHandle ON_COMPLETION;

    static {
        try {
            ON_COMPLETION = MethodHandles.lookup()
                    .findVirtual(
                            Completions.class,
                            "onCompletion",
                            MethodType.methodType(
                                    void.class,
                                    int.class,
                                    MemorySegment.class,
                                    MemorySegment.class,
                                    MemorySegment.class));
        } catch (ReflectiveOperationException e) {
            throw new ExceptionInInitializerError(e);
        }
    }

    private final ConcurrentHashMap<Long, Pending<?>> pending = new ConcurrentHashMap<>();
    private final AtomicLong nextId = new AtomicLong(1);
    private final MemorySegment callback;

    /**
     * @param arena the lifetime the completion stub belongs to — the client's, so it outlives every
     *     operation the client can still be called back about
     */
    public Completions(Arena arena) {
        this.callback = Upcalls.stub(ON_COMPLETION.bindTo(this), Ffi.Callbacks.COMPLETION, arena);
    }

    /**
     * Registers an operation and returns what to hand the FFI.
     *
     * @param operation which call this is, for the error it may produce
     * @param decoder how to read a successful payload
     * @param future the future to settle
     * @return the callback and userdata for the native call
     */
    public <T> Ticket register(String operation, Decoder<T> decoder, CompletableFuture<T> future) {
        long id = nextId.getAndIncrement();
        pending.put(id, new Pending<>(operation, decoder, future));
        return new Ticket(callback, MemorySegment.ofAddress(id));
    }

    /** How many operations are still waiting. The endurance suite watches this for a leak. */
    public int pendingCount() {
        return pending.size();
    }

    /**
     * Settles everything still pending, and stops finding anything afterwards.
     *
     * <p>Called when the client closes. Leaving an operation out would leave its caller waiting for
     * the life of the process, which is the other half of the same bug as freeing one too early.
     *
     * @param why what to fail them with
     */
    public void settleAll(ReactorException why) {
        List<Pending<?>> abandoned = new ArrayList<>(pending.values());
        pending.clear();
        for (Pending<?> one : abandoned) {
            one.future().completeExceptionally(why);
        }
    }

    private void onCompletion(int ok, MemorySegment resultJson, MemorySegment errorJson, MemorySegment userdata) {
        Pending<?> operation = pending.remove(userdata.address());
        if (operation == null) {
            // Already settled — a close that beat the completion, or a completion arriving twice.
            // Nothing to do, and nothing to complain about.
            return;
        }
        settle(operation, ok != 0, NativeStrings.borrow(resultJson), NativeStrings.borrow(errorJson));
    }

    private static <T> void settle(
            Pending<T> operation, boolean ok, @Nullable String resultJson, @Nullable String errorJson) {
        if (!ok) {
            operation.future().completeExceptionally(ErrorPayloads.parse(errorJson, operation.operation()));
            return;
        }
        T value;
        try {
            // Decoded first, claimed second. The other order cannot report a bad payload: the
            // future is already done by the time the conversion throws.
            value = operation.decoder().decode(resultJson);
        } catch (RuntimeException decodeFailed) {
            operation
                    .future()
                    .completeExceptionally(ReactorException.of(
                            inc.reactor.sdk.ErrorCode.DECODE_FAILED.code(),
                            "the reply to " + operation.operation() + " could not be understood: "
                                    + decodeFailed.getMessage(),
                            null,
                            operation.operation(),
                            null));
            return;
        }
        operation.future().complete(value);
    }
}

File: sdks/android/reactor-sdk-android/src/main/kotlin/inc/reactor/sdk/android/internal/Completions.kt
package inc.reactor.sdk.android.internal

import inc.reactor.sdk.android.ErrorCode
import kotlinx.coroutines.CompletableDeferred
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/**
 * One async FFI operation, bridged to a `suspend` function.
 *
 * Three rules, each of which the boundary punishes differently when broken.
 *
 * **Decode before claiming the promise.** A future settles once. If settling marks the operation
 * done and *then* converts the payload, a field of the wrong type throws where nothing can be
 * settled any more — the fallback meant to fail the call finds it already claimed, and the caller
 * gets a hang instead of the typed error this SDK documents. So [settle] converts first and
 * completes second.
 *
 * **A successful completion whose payload will not parse is a decode failure, not `{}`.**
 * Substituting an empty object makes a schema request answer with a schema declaring nothing,
 * which no caller can tell from a model that declares nothing. An *absent* payload — the ABI
 * sends null for a void operation — is a different answer, and does mean "nothing to report".
 *
 * **The ticket is independent of the client handle.** `reactor_fetch_jwt` takes no handle at all,
 * and `reactor_download_clip` is documented as outliving the one it was given, so a completion
 * can arrive after the client is gone. The registry therefore outlives any single client, and
 * [abandonAll] settles waiters rather than dropping them — a caller left holding an unresolved
 * promise waits for the life of the process.
 */
internal object Completions {
    private val nextTicket = AtomicLong(1)
    private val pending = ConcurrentHashMap<Long, Pending<*>>()

    private class Pending<T>(
        val operation: String,
        val decode: (String?) -> T,
        val deferred: CompletableDeferred<T>,
    )

    /**
     * Settle a completion from the native trampoline.
     *
     * `@JvmStatic` because `client.cpp` resolves it with `GetStaticMethodID`: a Kotlin `object`'s
     * members are instance methods on INSTANCE by default, and the bridge would have to carry a
     * reference to that instance for no reason. The ticket is all the native side holds.
     */
    @JvmStatic
    fun settleFromNative(
        ticket: Long,
        ok: Boolean,
        resultJson: String?,
        errorJson: String?,
    ) {
        settle(ticket, ok, resultJson, errorJson)
    }

    /** How many completions are outstanding. Zero at rest; the endurance suite watches it. */
    val pendingCount: Int
        get() = pending.size

    /**
     * Register a completion and hand its ticket to [start].
     *
     * [decode] runs on the native completion thread, before the promise is claimed. It receives
     * the raw `result_json`, which is null when the operation has nothing to report.
     */
    suspend fun <T> await(
        operation: String,
        decode: (String?) -> T,
        start: (ticket: Long) -> Unit,
    ): T {
        val ticket = nextTicket.getAndIncrement()
        val deferred = CompletableDeferred<T>()
        pending[ticket] = Pending(operation, decode, deferred)
        try {
            start(ticket)
        } catch (t: Throwable) {
            // The operation never started, so no completion will ever arrive for this ticket.
            pending.remove(ticket)
            throw t
        }
        return try {
            deferred.await()
        } finally {
            // On cancellation the ticket is dropped here, but the *native* operation keeps
            // running — nothing in this ABI can cancel one. A late completion for a ticket that
            // is gone is dropped by settle(), which is the correct outcome and not a leak: the
            // native side owns its own resources.
            pending.remove(ticket)
        }
    }

    /**
     * Settle a completion. Called from the native completion thread.
     *
     * Returns silently for an unknown ticket, which is the ordinary case after cancellation.
     */
    fun settle(
        ticket: Long,
        ok: Boolean,
        resultJson: String?,
        errorJson: String?,
    ) {
        @Suppress("UNCHECKED_CAST")
        val entry = pending.remove(ticket) as? Pending<Any?> ?: return

        if (!ok) {
            entry.deferred.completeExceptionally(
                ErrorPayloads.toException(errorJson, fallbackOperation = entry.operation),
            )
            return
        }

        // Decode first. Claiming the promise and then converting is the shape that produces a
        // hang rather than an error.
        val decoded =
            try {
                entry.decode(resultJson)
            } catch (e: Throwable) {
                entry.deferred.completeExceptionally(decodeFailure(entry.operation, resultJson, e))
                return
            }
        entry.deferred.complete(decoded)
    }

    /**
     * Settle every outstanding completion, because the client is going away.
     *
     * Says what actually happened: the native operation may well still be running — a download
     * whose client was destroyed is still downloading — and telling the caller the file may yet
     * arrive is worth more than "aborted".
     */
    fun abandonAll(reason: String) {
        val snapshot = pending.keys.toList()
        for (ticket in snapshot) {
            @Suppress("UNCHECKED_CAST")
            val entry = pending.remove(ticket) as? Pending<Any?> ?: continue
            entry.deferred.completeExceptionally(
                ErrorCode.toException(
                    wire = "ABORTED",
                    message = reason,
                    operation = entry.operation,
                ),
            )
        }
    }

    private fun decodeFailure(
        operation: String,
        payload: String?,
        cause: Throwable,
    ) = ErrorCode.toException(
        wire = "DECODE_FAILED",
        message = "The $operation completion succeeded but its payload did not decode: $payload",
        operation = operation,
        cause = cause,
    )

    /**
     * A duration a caller gave us, as milliseconds the ABI will accept.
     *
     * Numbers cross this boundary as whatever the host had, and a NaN or an infinity turned into
     * a duration panics *inside a detached native task* — which drops the completion instead of
     * firing it, leaving the binding waiting for a callback that can no longer come. Checked here
     * so the refusal is an ordinary exception on the calling thread.
     *
     * Negative and infinite both mean "no bound"; a NaN is a caller bug; a finite value too large
     * for the ABI saturates rather than wrapping.
     */
    fun timeoutMillis(seconds: Double?): Long {
        if (seconds == null) return 0L
        if (seconds.isNaN()) {
            throw ErrorCode.toException(
                wire = "BAD_REQUEST",
                message = "A timeout of NaN is not a duration",
            )
        }
        if (seconds.isInfinite() || seconds < 0.0) return 0L
        val millis = seconds * 1000.0
        return if (millis >= Long.MAX_VALUE.toDouble()) Long.MAX_VALUE else millis.toLong()
    }
}
